volmigrate/volmigrate.go

```diff
@@ -0,0 +1,170 @@
+	"github.com/contiv/volplugin/config"
+	"github.com/contiv/volplugin/volmigrate/backend"
+	"github.com/contiv/volplugin/volmigrate/backend/etcd2"
+)
+
+// -------------------------------------------------------------------------------------------------
+
+// NOTE: These three functions are copied from volcli.go
+//       Rather than creating a one-off package for them, they are intentionally directly copied.
+//       If another app like volcli/volmigrate is added in the future, we should extract these
+//       to a separate package.
+
+func errExit(ctx *cli.Context, err error, help bool) {
+	fmt.Fprintf(os.Stderr, "\nError: %v\n\n", err)
+	if help {
+		cli.ShowAppHelp(ctx)
+	}
+	os.Exit(1)
+}
+
+func execCliAndExit(ctx *cli.Context, f func(ctx *cli.Context) (bool, error)) {
+	if showHelp, err := f(ctx); err != nil {
+		errExit(ctx, err, showHelp)
+	}
+}
+
+func errorInvalidArgCount(rcvd, exptd int, args []string) error {
+	return errored.Errorf("Invalid number of arguments: expected %d but received %d %v", exptd, rcvd, args)
+}
+
+// -------------------------------------------------------------------------------------------------
+
+func latestMigrationVersion() int64 {
+	return availableMigrations[len(availableMigrations)-1].Version
+}
+
+func promptBeforeRunning(ctx *cli.Context, msg string) {
+	if ctx.GlobalBool("silent") {
+		return
+	}
+
+	fmt.Println(msg)
+	fmt.Println("Are you sure you want to continue? y/n")
+
+	reader := bufio.NewReader(os.Stdin)
+	text, err := reader.ReadString('\n')
+	if err != nil {
+		log.Fatalf("Failed to read input from stdin: %s", err)
+	}
+
+	text = strings.ToLower(text)
+	text = text[:len(text)-1] // strip newline
+
+	if text != "yes" && text != "y" {
+		fmt.Println("Aborting.")
+		os.Exit(1)
+	}
+}
+
+func List(ctx *cli.Context) {
+	execCliAndExit(ctx, list)
+}
+
+func list(ctx *cli.Context) (bool, error) {
+	if len(ctx.Args()) != 0 {
+		return true, errorInvalidArgCount(len(ctx.Args()), 0, ctx.Args())
+	}
+
+	for _, m := range availableMigrations {
+		fmt.Printf("%d - %s\n", m.Version, m.Description)
+	}
+
+	return false, nil
+}
+
+func Run(ctx *cli.Context) {
+	execCliAndExit(ctx, run)
+}
+
+func run(ctx *cli.Context) (bool, error) {
+	e := etcd2.New(ctx.GlobalString("prefix"), ctx.GlobalStringSlice("etcd"))
+
+	if len(ctx.Args()) > 0 {
+		i, err := strconv.Atoi(ctx.Args()[0])
+		if err != nil {
+			return true, errored.Errorf("Invalid migration version: %v", ctx.Args()[0])
+		}
+
+		promptBeforeRunning(ctx, fmt.Sprintf("You have requested to run all pending migrations up to and including #%d.", i))
+		return runMigrationsUpTo(e, int64(i))
+	}
+
+	promptBeforeRunning(ctx, "You have requested to run all pending migrations.")
+	return runAllMigrations(e)
+}
+
+func runAllMigrations(b backend.Backend) (bool, error) {
+	return runMigrationsUpTo(b, latestMigrationVersion())
+}
+
+func findMigrationByVersion(cfg *config.Client, version int64) (*Migration, bool) {
+	for _, m := range availableMigrations {
```

Comment:
Good point, but that also assumes that there are no gaps in the migration version sequence.   Currently, `registerMigration()` *would* allow that behavior.

The reason I made the version ID explicit (i.e., an argument to `registerMigration()`) is because having it auto-increment when a new migration is registered makes migration IDs position-dependent in the source.  It would also make it harder to find a specific migration by ID when looking at the source code.

As a middle ground, I can ensure that when a migration is registered it has the expected ID (the next in sequence) and abort otherwise.  Then this simpler code you've written would work and there's no danger of position-dependent migration IDs or sequence gaps.

Thanks, I will do this.